packages/core/controls.mjs

```diff
@@ -4,7 +4,8 @@ Copyright (C) 2022 Strudel contributors - see <https://github.com/tidalcycles/st
 This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details. You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
 
-import { Pattern, sequence } from './pattern.mjs';
```

Comment:
wouldn't it be simpler if we kept those export as they are and only export an additional `pattern` object? this way we can still have simpler imports (left side), while having a monster object to use as a scope.

Reply:
Hm, but then there's no clear reasoning for why a function is exported directly, or only available in the pattern object, and it might appear and disappear based on how its implemented internally.

Reply:
hm yes all the dynamically generated functions would only appear in the big object, the rest could be exported separately (in addition to being in the big object). this just simplifies importing 90% of all things, because you don't have to always do `const { x, y, z } = core`. also there won't be breaking changes everywhere. We could even go so far to write a script that exports all the matrix functions in the root scope..

Reply:
Hm I think I'd rather just scrap all this and generate sourcecode then. We just do everything with methods in pattern.mjs, list or mark the functions we want to promote to curried toplevel functions, and generate that as a separate source file called something like `pattern-static.mjs`. We then add a build script like we do in the mininotation package.

Reply:
There'd then be a question about whether to generate controls.mjs to be exported directly as well. It's quite nice to have them organised a bit separately, but if they all end up in the same scope anyway.. To be consistent, best to either put everything in one object, or have everything directly exported?

Reply:
it would still be practical to have a non generated big object with all things that should go into the eval scope. for all internal imports this should be fine. i see the source code generation more as a way to provide a simpler import api for library users

Reply:
I think almost all the non-methods should be either dynamically or statically generated. Otherwise we get this sort of thing which difficult to maintain:
https://github.com/tidalcycles/strudel/blob/485a96962ad4a8057a8387f14ae1fa42755e9873/packages/core/pattern.mjs#L1822-L1852

One possible advantage of statically generated rather than dynamically generated code is that we can generate documentation along with it, so that conventional documentation tools work.

Reply:
> I think almost all the non-methods should be either dynamically or statically generated. Otherwise we get this sort of thing which difficult to maintain:

agreed

> One possible advantage of statically generated rather than dynamically generated code is that we can generate documentation along with it, so that conventional documentation tools work.

jsdoc also works in thin air, when used with @name and @memberof. So this should be no limiting factor

Reply:
but i'd say static written code is much easier to read

Reply:
If the main argument is that a monster object is a breaking change, I think that's fine. Having a single, consistent way to export stuff means less breaking in the future.. and this PR already fixes all the internal breaks I could find.

For the memory usage argument, I don't know how javascript deals with memory but guess we could try it..

Reply:
> but i'd say static written code is much easier to read

The problem is there are inconsistencies and a lot of missing toplevel functions. I don't really want to manually fix this as it'd be an ongoing headache to keep it up-to-date.

Unless we switched things around and wrote everything as a toplevel module function, with a decorator (via babel?) to add it to the class as a method.

Reply:
talking about memory i meant tree shaking, which means only the things that are imported are used. Breaking changes is one argument, but the main argument is that default exporting big objects is generally considered to be an "antipattern".. it's more difficult to auto import stuff + you always have to write an extra statement by hand.. i heavily use code completion which just automatically imports the stuff i am using, which does not work this way.. some written words about this: https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

Reply:
i still use default exports btw but only when a file exports a like function / value etc..

Reply:
Here's the Tidal equivalent of controls.mjs: https://github.com/tidalcycles/Tidal/blob/main/src/Sound/Tidal/Control.hs

It's autogenerated and reads the same as when it was handmade.. just more consistent and with fewer bugs https://github.com/tidalcycles/Tidal/blob/main/src/Sound/Tidal/Control.hs

Reply:
> The problem is there are inconsistencies and a lot of missing toplevel functions. I don't really want to manually fix this as it'd be an ongoing headache to keep it up-to-date.

yes, I also want this to be autogenerated (I mean autogenerated on an object, not autogenerated with a script that outputs code)

> Unless we switched things around and wrote everything as a toplevel module function, with a decorator (via babel?) to curry it and add it to the class as a method.

This is (without huge hassle) only practical as a build step for distribution. Inside the code base, this should be pretty hard to do, as we have to integrate it into the dev server / hot reloading etc.. imo not worth it just to be able to import things from dynamically generated objects in one line less.

Reply:
Ok I'm happy to accept that the monster object export isn't ideal.

That's a shame about lack of decorator but I guess that's only syntactic sugar anyway

```js
export class Pattern {
  // .. bare minimum goes here
}

export const every = _register(function(n, func, pat) {
    // ... definition of every
  }
);
```

Where `_register` adds the method to the class, and returns a curried version.. hopefully sorting out the composition stuff as well. Then we get both the exports and the class of methods.

We have to define each of the operators by hand but that's fine now that it's of the form add.mix rather than addMix.

Reply:
that looks like it could solve this mess..

Reply:
we need to pass the function name i think (to set `Pattern.prototype[name]` programmatically) 

```js
_register('every', function(n, func, pat) {
    // ... definition of every
  }
);
```

Reply:
I'll have to sleep over this...

Reply:
Ah yes.. 

With this approach we could also add the ability to 'bless' anything into being a pattern, not just the Pattern class.

Will have a think as well. I can extract the matrix howWhat -> how.what change into a separate PR at some point..

Reply:
Ah yes.. 

With this approach we could also add the ability to 'bless' anything into being a pattern, not just the Pattern class.

Will have a think as well. I can extract the matrix howWhat -> how.what change into a separate PR at some point..